### ISSUE TITLE
Is this even necessary?

### ISSUE THREAD

**Issue by doodledood** [cyclejs/cycle-http-driver#42]
I am trying to get a broader understanding of Cycle JS's architecture and the reasoning behind it.

My question is this:
Do we even need to have super abstract drivers like the HTTP driver and the LocalStorage driver?
Why can't I just create an "API" driver specific to my app and avoid all the inconvenience that comes with using a wrapper around some arbitrary library? (For example superagent in the HTTP driver)
Isn't the whole point of having drivers to abstract away the "how" from the "what"?

If I just create an "API" driver and simply deliver "commands" to it via a sink and receive results via a source, I don't need to worry about the finer details (like filtering a GET request by the address) - if I for example, currently use local storage to retrieve my data and I now want to use an external server, I will need to change code in many places instead of only a single place (the API driver). With a specific API driver one I can do just that - which supports the Open/Closed principle.

Basically using an API driver looks like a mini-flux inside Cycle JS where the API sink is dispatching an action and the API driver being a reducer/saga that contains all of the side effects in the system.

Am I missing something? What are your thoughts about this?

**Comment by Frikki** [comment]
I wholeheartedly agree.

**Comment by axefrog** [comment]
Last week I delivered a small project for a client and this is exactly what I did. I basically built a "state" driver which took state change requests (sink) and emitted state changes (source). Whilst the general idea behind drivers is that they give you a tool for separating the side effects from your application code, there's no reason you can't take it a step further and employ a functional approach to business logic within your driver (where possible) and perform your own separation between side-effectful code and business logic.

**Comment by whitecolor** [comment]
@doodledood
Yes you are correct in some way. But. First, this driver gives an a good example and starting point for build other async request/response drivers. The second? the way you describe getting state to the app itself requires logic (get from remote, take from/to local storage) - this means that this logic should be implemented some where so your data management itself become an app (it probably a cycle app). The result of working of this app can be application state that you provide to another cycle function. This approach can even be impleneted as separate cyclejs apps which have to connected via some driver. 

So your "mini-flux" driver is going to include capabilities of other basic drivers, but if there is logic inside, it would be better to implement it internaly as cycle app itself based on basic drivers, or you end up with driver that do much imperative logic inside.

You're just talkin about some higher level achitecture pattern that can be beautifully implement using cycle approach.

**Comment by staltz** [comment]
> Is this even necessary?

What does "necessary" mean? If you mean "necessary for getting things done", then you don't need Cycle.js either. jQuery is enough. Not even that is necessary, [you might not need jQuery](http://youmightnotneedjquery.com/), there's even nice ajax built-in (`Fetch`).

I understand that this driver is "necessary" *to separate logic from effects*. 

> So your "mini-flux" driver is going to include capabilities of other basic drivers, but if there is logic inside

Precisely. That's what I was going to say. If your driver is specific to your application and cannot be trivially reused in another unrelated application, then it's a symptom that your driver contains business logic. In that case you don't have anymore logic / effects separation.

> If I just create an "API" driver and simply deliver "commands" to it via a sink and receive results via a source

Instead of a driver, what you could also implement is your ideal API as a post-processing step right before returning sinks from `main`, then a pre-processing step right after receiving `sources`. Then you're essentially abstracting away this HTTP Driver low-level details and you're using a nice API but you still get "all logic inside main".

**Comment by doodledood** [comment]
@staltz I agree with you about the pre/post processing steps instead of an actual driver. It does seem like a more elegant solution that somewhat keeps the same level of abstraction in your main while keeping side effects away.

Thank you all for answering.

**Comment by staltz** [comment]
Alright, I'll close this issue, we can reopen if necessary later.

**Comment by SkaterDad** [comment]
@staltz Can you elaborate on what you mean by `ReadOnlyHTTPDriver`?

I do really like the idea of being able to issue `get` or `post` requests without having to pipe the the `http` sink all the way back to the primary `main()` function.  Or have I misinterpreted?

**Comment by staltz** [comment]
@SkaterDad this https://gitter.im/cyclejs/cycle-core?at=568306453acb611716002b80

**Comment by whitecolor** [comment]
@staltz 
> One solid idea is a ReadOnlyHTTPDriver, where you can do something like sources.HTTP.get('http://google.com')
Who and when desides when to make a request in this case?

**Comment by staltz** [comment]
The resource is assumed to exist and it's assumed to be static, just like with the local storage driver. The request happens when the observable which that returns is used by others. In practice it would happen on app startup, but could happen later.

**Comment by whitecolor** [comment]
Ok, so its for simple static (single) use case.

**Comment by staltz** [comment]
Yes, only. For anything more complicated, I think Falcor may be a better abstraction.

**Comment by whitecolor** [comment]
@staltz 
So such capablity could be added even to exsisting HTTP driver to allow any kind of "one time" requests?
```
HTTP.request({url:..., method: 'put'}).
```
Is anyting bad in such pattern?

**Comment by SkaterDad** [comment]
@whitecolor I think in many cases, these _one time use_ HTTP requests will be a big improvement.   Particularly for GET requests and simple POST requests.

Brainstorming, one downside to issuing a 'put' or 'post' with the _one-time use_ HTTP request could be that other parts of your app cannot intercept the reply from the server.

One example of this is if you had a routing component which needs to redirect after specific POST requests are successful.  Of course we can't see true redirects since the browser handles them, but lets say we're decorating our server responses with a `redirectTo` field.

With the current driver, you can have a single observable inspecting all of the responses, looking for that field, and sending a new value to the history driver.

With the component-specific HTTP post, your component would have to pass along the new url manually.

**Comment by HighOnDrive** [comment]
It would be nice to see Falcor from a Cycle use case. Otherwise, I think using RethinkDB to init/restore at app startup and changefeed a KV map of atoms and states will work just fine. Then just keep receiving changefeeds as atoms and states are modified via a driver that can be passed around as a source throughout the app. Why use one shot HTTP requests when you can have changefeeds?
